mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManagerMigration.java

```diff
@@ -0,0 +1,96 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CassandraPerUserMaxQuotaManagerMigration.class);
+    private final UsersRepository usersRepository;
+    private final DomainList domainDao;
+    private final MaxQuotaManager oldMaxQuotaManager;
+    private final MaxQuotaManager newMaxQuotaManager;
+    private UserQuotaRootResolver userQuotaRootResolver;
+
+    @Inject
+    public CassandraPerUserMaxQuotaManagerMigration(UsersRepository usersRepository, DomainList domainDao,
+                                                    @Named("old") MaxQuotaManager oldMaxQuotaManager, MaxQuotaManager newMaxQuotaManager,
+                                                    UserQuotaRootResolver userQuotaRootResolver) {
+        this.usersRepository = usersRepository;
+        this.domainDao = domainDao;
+        this.oldMaxQuotaManager = oldMaxQuotaManager;
+        this.newMaxQuotaManager = newMaxQuotaManager;
+        this.userQuotaRootResolver = userQuotaRootResolver;
+    }
+
+    @Override
+    public void apply() throws InterruptedException {
+        Mono.from(oldMaxQuotaManager.getGlobalMaxMessageReactive()).flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setGlobalMaxMessageReactive(quotaCountLimit))).block();
+        Mono.from(oldMaxQuotaManager.getGlobalMaxStorageReactive()).flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setGlobalMaxStorageReactive(quotaSizeLimit))).block();
+
+        Flux.fromIterable(getDomains())
+            .flatMap(domain -> Mono.from(oldMaxQuotaManager.getDomainMaxMessageReactive(domain))
+                .flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setDomainMaxMessageReactive(domain, quotaCountLimit)))
+                .then(Mono.from(oldMaxQuotaManager.getDomainMaxStorageReactive(domain))
+                    .flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setDomainMaxStorageReactive(domain, quotaSizeLimit)))))
+            .then()
+            .block();
+
+        Flux.from(usersRepository.listReactive())
+            .map(userQuotaRootResolver::forUser)
+            .flatMap(quotaRoot -> Mono.from(oldMaxQuotaManager.listMaxMessagesDetailsReactive(quotaRoot))
+                .map(map -> map.get(Quota.Scope.User))
+                .flatMap(quotaCountLimit -> Mono.from(newMaxQuotaManager.setMaxMessageReactive(quotaRoot, quotaCountLimit)))
+                .then(Mono.from(oldMaxQuotaManager.listMaxStorageDetailsReactive(quotaRoot))
+                    .map(map -> map.get(Quota.Scope.User))
+                    .flatMap(quotaSizeLimit -> Mono.from(newMaxQuotaManager.setMaxStorageReactive(quotaRoot, quotaSizeLimit)))))
+            .then()
+            .block();
+    }
+
+    private List<Domain> getDomains() {
+        List<Domain> domains = new ArrayList<>();
+        try {
+            domains = domainDao.getDomains();
+        } catch (DomainListException ex) {
+            LOGGER.warn("Could not get domain list", ex);
+        }
+        return domains;
+    }
```

Comment:
Why?

Reply:
domainDao.getDomains() throw DomainListException. In case the exception occur, log an warning and then continue migration by returning empty list.

Reply:
Then you are lying to the operator who thought the migration to be done while it is not the case. 

It is better to thorw here!